buildscripts/set_tensorflow_nvidia_bazelrc.sh

```diff
@@ -45,8 +45,8 @@ build --action_env TF_CUDA_VERSION="${cudatoolkit%.*}"
 build --action_env TF_CUDNN_VERSION="${cudnn:0:1}" #First digit only
 build --action_env TF_TENSORRT_VERSION="${tensorrt:0:1}"
 build --action_env TF_NCCL_VERSION="${nccl:0:1}"
-build --action_env TF_CUDA_PATHS="$CUDA_TOOLKIT_PATH"
-build --action_env CUDA_TOOLKIT_PATH="$CUDA_TOOLKIT_PATH"
+build --action_env TF_CUDA_PATHS="$OPEN_CE_CUDA_HOME,$CUDA_TOOLKIT_PATH"
```

Comment:
Should the order here be reversed so that it will use libraries that exist in `$CUDA_TOOLKIT_PATH` first? Otherwise it seems like it would always be skipping the CUDA libraries in the CUDA environment.

Reply:
I was trying determine the behavior here as well. I was worried it would do some cursory checks on the paths given here, find some token files, and call a path good and stop looking or caring about other paths in the list.

What we want is `nvcc` being found in the CONDA environment, but any libraries needed by the linker would be found in the OPEN_CE_CUDA_HOME. I think the fact that its finding TensorRT here proves it holds on to both paths throughout. 

So the next thought is the ordering. Presumably it searched the paths sequentially, so it very well may be finding nvcc in OPEN_CE_CUDA_HOME first.

Reply:
I think we want it to find the libraries within the `conda` environment as well. The libraries should be installed within the conda environment by the `cudatoolkit` package. This is what we want to link against. (If there are any static libraries that aren't part of the `cudatoolkit` package, then we should look in `OPEN_CE_CUDA_HOME`).

Reply:
Currently, we aren't requiring the `cudatoolkit` package at build time in the recipes (and its not in nvcc as a runtime dep either).
The nvcc package points to OPEN_CE_CUDA_HOME when running `nvcc`. And it finds everything it needs to build there. The `cudatoolkit` package is used at runtime only.
Perhaps we could change that, but I don't think that simplifies anything. And it could perhaps cause issues if the cudatoolkit package differs slightly than the what is installed on the host (but is at the same cuda level) for whatever reason.

Reply:
The other thing I noticed is the `CUDA_TOOLKIT_PATH` is still pointing to `$PREFIX` not `CONDA_PREFIX`. It does seem to work this way.

Reply:
Changing the order here is perhaps significant. I get a build fail

```
ERROR: An error occurred during the fetch of repository 'local_config_cuda':
   Traceback (most recent call last):
	File "/opt/conda/conda-bld/tensorflow-base_1602016537047/work/third_party/gpus/cuda_configure.bzl", line 1400
		_create_local_cuda_repository(<1 more arguments>)
	File "/opt/conda/conda-bld/tensorflow-base_1602016537047/work/third_party/gpus/cuda_configure.bzl", line 1080, in _create_local_cuda_repository
		_find_libs(repository_ctx, <2 more arguments>)
	File "/opt/conda/conda-bld/tensorflow-base_1602016537047/work/third_party/gpus/cuda_configure.bzl", line 627, in _find_libs
		_check_cuda_libs(repository_ctx, <2 more arguments>)
	File "/opt/conda/conda-bld/tensorflow-base_1602016537047/work/third_party/gpus/cuda_configure.bzl", line 529, in _check_cuda_libs
		execute(repository_ctx, <1 more arguments>)
	File "/opt/conda/conda-bld/tensorflow-base_1602016537047/work/third_party/remote_config/common.bzl", line 208, in execute
		fail(<1 more arguments>)
Repository command failed
No library found under: /opt/conda/conda-bld/tensorflow-base_1602016537047/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/lib/stubs/libcuda.so
```

Reply:
The `nvcc` package will be under $CONDA_PREFIX if `compiler('cuda')` is listed under the `build` category in the meta.yaml file. This is probably where it should be moved and then change CUDA_TOOLKIT_PATH to CONDA_PREFIX.

Reply:
I've tried moving compiler('cuda') to build section and then changing CUDATOOLKIT to CONDA_PREFIX but it fails to find Tensorrt stuff. So, have to keep both PREFIX as well as CONDA_PREFIX.

Reply:
So, what works is -
- having compiler('cuda') in build section instead of host section
- TF_CUDA_PATHS="$OPEN_CE_CUDA_HOME,$CONDA_PREFIX,$PREFIX"  -----> PREFIX is needed for TRT to be located.

Reply:
Regarding changing order of paths - 
I tried putting OPEN_CE_CUDA_HOME at the end, but it fails because TF expects nvcc binary path and nvvm paths should be from same directory (check is https://github.com/tensorflow/tensorflow/blob/v2.3.0/third_party/gpus/find_cuda_config.py#L285). And then the 1st path is used as cudatoolkit install directory to locate all the other binaries like nvlink, fatbinary, bin2c, etc., which won't be present in CONDA_PREFIX being 1st path.
I tried patching find_cuda_config.py from tensorflow to handle above problems, keeping OPEN_CE_CUDA_HOME at the end. Manually running find_cuda_config.py script with these path ordering I get following, which is what we want -
```
cuda_binary_dir: /opt/conda/conda-bld/tensorflow-base_1602073149565/_build_env/bin
cuda_include_dir: /opt/conda/conda-bld/tensorflow-base_1602073149565/_build_env/include
cuda_library_dir: /opt/conda/conda-bld/tensorflow-base_1602073149565/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/lib
cuda_toolkit_path: /usr/local/cuda-10.2
cuda_version: 10.2
```

Reply:
@jayfurmanek - I think in nvcc package, we might have to create symlink for libcuda.so. We are creating one for libcuda.so.1 but TF build is looking for libcuda.so.

Reply:
yeah. TF *should* be looking for the `.1` if it was proper, but a lot of things in TF are not exactly proper.
I don't see any hard in adding another symlink there.

Reply:
Also, find_cuda_config.py searches for libcudart.so and returns its path as `cuda_library_dir` where libcuda.so is searched. In our case, libcudart.so is found in $PREFIX (installed by cudatoolkit). So, we'll also have to have cuda compiler in host section so that it puts libcuda.so in $PREFIX and then use that $PREFIX in CUDA_TOOLKIT_PATHS instead of $CONDA_PREFIX.

Reply:
If adding a link, you should probably have it just point to the libcuda.so.1 symlink instead of the one at OPEN_CE_CUDA_HOME.

Reply:
Next error is coming for libcudart_static.a which is not present in our $PREFIX or $CONDA_PREFIX. It's in $OPEN_CE_CUDA_HOME. I'll check TF how to fix static libs paths.

Reply:
yeah NCCL links to `libcudart_static.a` as well. In that environment there luckily was a `CUDA_LIBS` arg that get passed to the `-L` in the linker (which is the compiler called directly, not through nvcc)

For Tensorflow, its less clear, but there is only the CUDA_TOOLKIT_PATH and TF_CUDA_PATHS to play with.

Reply:
As I mention above, the `cudatoolkit` package shouldn't be installed at build time though. We just have `nvcc` meta/wrapper and the local CUDA installation.

Reply:
cudatoolkit package comes via tensorrt probably.

Reply:
ah..right. And NCCL, cuDNN are installed as well which drag `cudatoolkit` in.
That's all unavoidable, so we'll have to deal with `cudatoolkit` being there afterall.

Reply:
I think it appears weird that shared libs are being used from one place and static libs from another. Problem with TF is although we provide a list of paths in CUDA_TOOLKIT_PATH, it finally applies its own logic and use just one directory for libraries where libcudart.so is found.